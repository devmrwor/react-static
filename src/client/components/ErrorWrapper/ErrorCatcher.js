import React from 'react'
import PropTypes from 'prop-types'
import ErrorUI from './ErrorUI'

export default class ErrorCatcher extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  }

  state = {
    error: null,
    errorInfo: null,
  }

  componentDidCatch (error, errorInfo) {
    // Catch errors in any child components and re-renders with an error message
    this.setState({
      error,
      errorInfo,
    })
  }

  render () {
    const { error } = this.state

    // Fallback UI if an error occurs
    if (error) {
      return <ErrorUI {...this.state} />
    }

    return React.Children.only(this.props.children)
  }
}
