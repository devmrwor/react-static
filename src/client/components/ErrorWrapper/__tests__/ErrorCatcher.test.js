import React from 'react'
import { shallow, mount } from 'enzyme'
import ErrorCatcher from '../ErrorCatcher'

describe('ErrorCatcher', () => {
  test('by default should render children', () => {
    const errorCatcher = shallow(
      <ErrorCatcher>
        <span>hello</span>
      </ErrorCatcher>
    )

    expect(errorCatcher).toMatchSnapshot()
  })

  describe('when children throw an error', () => {
    it('should catch errors with componentDidCatch', () => {
      const ChildComponentWithError = () => {
        throw new Error('Error thrown from problem child')
      }

      jest.spyOn(ErrorCatcher.prototype, 'componentDidCatch')

      const errorCatcher = mount(
        <ErrorCatcher>
          <ChildComponentWithError />
        </ErrorCatcher>
      )

      expect(ErrorCatcher.prototype.componentDidCatch).toHaveBeenCalled()
      expect(errorCatcher.state('error')).toEqual(new Error('Error thrown from problem child'))
      expect(errorCatcher.state('errorInfo')).toMatchSnapshot()
      expect(errorCatcher).toMatchSnapshot()
    })
  })
})
