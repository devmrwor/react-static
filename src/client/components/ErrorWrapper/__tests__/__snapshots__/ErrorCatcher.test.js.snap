// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorCatcher by default should render children 1`] = `
<span>
  hello
</span>
`;

exports[`ErrorCatcher when children throw an error should catch errors with componentDidCatch 1`] = `
Object {
  "componentStack": "
    in ChildComponentWithError (at ErrorCatcher.test.js:26)
    in ErrorCatcher (created by WrapperComponent)
    in WrapperComponent",
}
`;

exports[`ErrorCatcher when children throw an error should catch errors with componentDidCatch 2`] = `
<ErrorCatcher>
  <ErrorUI
    error={[Error: Error thrown from problem child]}
    errorInfo={
      Object {
        "componentStack": "
    in ChildComponentWithError (at ErrorCatcher.test.js:26)
    in ErrorCatcher (created by WrapperComponent)
    in WrapperComponent",
      }
    }
  >
    <div
      style={
        Object {
          "background": "rgba(0,0,0,0.05)",
          "margin": "1rem",
          "padding": "1rem",
        }
      }
    >
      <h2>
        Oh-no! Somethingâ€™s gone wrong!
      </h2>
      <pre
        style={
          Object {
            "color": "red",
            "whiteSpace": "normal",
          }
        }
      >
        <code>
          Error: Error thrown from problem child
        </code>
      </pre>
      <h3>
        This error occurred here:
      </h3>
      <pre
        style={
          Object {
            "color": "red",
            "overflow": "auto",
          }
        }
      >
        <code>
          
    in ChildComponentWithError (at ErrorCatcher.test.js:26)
    in ErrorCatcher (created by WrapperComponent)
    in WrapperComponent
        </code>
      </pre>
      <p>
        For more information, please see the console.
      </p>
    </div>
  </ErrorUI>
</ErrorCatcher>
`;
